### PR TITLE
EJB App Packaging for Mac & Linux

### PR DESCRIPTION
**Pull request by vitvakatu**
## Overview

---
See: https://jira.bf.local/browse/ECR-XYZ

### Definition of Done

- [ ] There are no TODOs left in the code
- [ ] Change is covered by automated [tests](https://github.com/exonum/exonum-java-binding/blob/master/CONTRIBUTING.md#tests)
- [ ] The [coding guidelines](https://github.com/exonum/exonum-java-binding/blob/master/CONTRIBUTING.md#the-code-style) are followed
- [ ] Public API has Javadoc
- [ ] Method preconditions are checked and documented in the Javadoc of the method
- [ ] Changelog is updated if needed (in case of notable or breaking changes)
- [ ] The continuous integration build passes

### CODE REVIEW

File: exonum-java-binding/core/rust/ejb-app/TUTORIAL.md [hunk 1]
```diff
@@ -4,7 +4,8 @@ This document describes how to configure and run an Exonum node with a Java serv
 ## Prerequisites
 
 Build an application following the instructions in [“How to Build”][how-to-build] section
-of the Contribution Guide.
+of the Contribution Guide. Unpack the zip archive from `exonum-java-binding/core/target` directory to some known
```
Comment (bullet-tooth) [inline exonum-java-binding/core/rust/ejb-app/TUTORIAL.md]: please add the command example, like `unzip exonum-java-binding/core/target/app.zip -d destination_folder`

**Comment by dmitry-timofeev** [comment]
@vitvakatu Why does it target `master`, including all the changes made so far? Can we review only the updated bits, targeting `ejb-app-configuration`? If you, however, pushed [these](https://github.com/exonum/exonum-java-binding/compare/0d192bf3de5123366e8ca0ce9550bcbb27abfec9...aae9b3695312c98acdfe5f7d2f464489254a1649) directly, please target a new branch off 0d192bf3de5123366e8ca0ce9550bcbb27abfec9.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -36,11 +38,24 @@ $ mvn install
 
 #### Building Java Binding App
 Run:
+
 ```$sh
-$ cd exonum-java-binding/core/rust/ejb-app
-$ cargo build
+$ cd exonum-java-binding
+$ ./package_app.sh
```
[the remark this answers is not in the document]
Comment (vitvakatu) [reply CONTRIBUTING.md]: It requires installed cmake (afaik, for generate-licenses.sh script), but it works on Mac, of course

[the remark this answers is not in the document]
Comment (vitvakatu) [reply CONTRIBUTING.md]: Yep, it looks like CONTRIBUTING.md needs a little patch to remove the line that it's only available on linux

File: exonum-java-binding/core/pom.xml [hunk 3]
```diff
@@ -26,6 +26,7 @@
     <rust.compiler.features>resource-manager</rust.compiler.features>
     <rust.itSubCrate>integration_tests</rust.itSubCrate>
     <build.nativeLibPath>${project.basedir}/rust/target/debug</build.nativeLibPath>
+    <packaging.basedir>${project.basedir}/rust/target/prepackage</packaging.basedir>
```
[the remark this answers is not in the document]
Comment (vitvakatu) [reply exonum-java-binding/core/pom.xml]: It collects some files needed on `package` maven phase.

File: exonum-java-binding/core/rust/ejb-app/TUTORIAL.md (hunk 1 above)
Comment (vitvakatu) [reply exonum-java-binding/core/rust/ejb-app/TUTORIAL.md]: What if user does not have `unzip` honestly I don't think it's necessary

**Comment by vitvakatu** [comment]
@dmitry-timofeev the reason is, `vitvakatu/ejb-app-configuration` is in sync with the latest exonum/master, but `exonum/ejb-app-configuration` is not (and includes some fixes after #732 reviews). So if I target `exonum/ejb-app-configuration` - we'll get huge amount of changes, including changes from master.

File: exonum-java-binding/tests_profile [hunk 4]
```diff
@@ -49,17 +49,14 @@ export JAVA_LIB_DIR="$(find ${JAVA_HOME} -type f -name libjvm.\* | xargs -n1 dir
...
 export RUST_COMPILER_VERSION="${RUST_COMPILER_VERSION:-1.32.0}"
 echo "RUST_COMPILER_VERSION: ${RUST_COMPILER_VERSION}"
 
 # Find the directory containing Rust libstd.
 export RUST_LIB_DIR=$(rustup run ${RUST_COMPILER_VERSION} rustc --print sysroot)/lib
 echo "RUST_LIB_DIR: ${RUST_LIB_DIR}"
 
-# Linking options with additional paths for the application and tests.
-# Note: `-C link-args` can be used only once, therefore, `-C link-arg` is used.
+# Linking options with additional paths for the application.
```
Comment (dmitry-timofeev) [inline exonum-java-binding/tests_profile]: For the application or for the tests ? We seem to set for the app in the packaging script?

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -36,11 +38,24 @@ $ mvn install
...
+in a single `zip` archive in `exonum-java-binding/core/target` directory.
+
+Before packaging, the script will also run all the tests to guarantee that the generated application
+will be valid. It may take a long time, so you can pass `--skip-tests` flag to skip this step:
+
+```$sh
+$ ./package_app.sh --skip-tests
+```
+
+**Automatic EJB App packaging is only supported on Linux platform.**
```
Comment (dmitry-timofeev) [inline CONTRIBUTING.md]: How about :green_apple: ?

File: exonum-java-binding/core/rust/ejb-app/TUTORIAL.md [hunk 6]
```diff
@@ -18,90 +19,79 @@ in the [user guide](https://exonum.com/doc/version/0.10/get-started/java-binding
...
-the Exonum Java Binding.
+Classpath is used to locate Java classes of your service and its dependencies.
 
 You may package your service in an Uber JAR using 
 the [Maven Shade Plugin](https://maven.apache.org/plugins/maven-shade-plugin/index.html)
-and pass a path to the service artefact during application configuration as `--ejb-classpath`
-parameter. Alternatively, you may assemble a classpath that includes the service and all of 
+and pass a path to the service artefact during application configuration as `--ejb-service-classpath`
+parameter. Alternatively, you may assemble a classpath that includes the path to service and all of 
 its dependencies and pass it instead.
```
[the remark this answers is not in the document]
Comment (vitvakatu) [reply exonum-java-binding/core/rust/ejb-app/TUTORIAL.md]: Fixed

File: exonum-java-binding/core/pom.xml [hunk 7]
```diff
@@ -342,6 +374,35 @@
...
+            <id>copy-core-jar</id>
+            <phase>install</phase>
+            <goals>
+              <goal>copy</goal>
+            </goals>
+            <configuration>
+              <outputDirectory>${project.basedir}/rust/target/debug/lib/java</outputDirectory>
+              <artifactItems>
+                <artifactItem>
+                  <groupId>com.exonum.binding</groupId>
```
Comment (dmitry-timofeev) [inline exonum-java-binding/core/pom.xml]: Here and elsewhere, use properties please: `${project.groupId}`, etc.

Also, why is this needed, if the code in 'packaging' does a similar thing?

File: exonum-java-binding/packaging/pom.xml [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+  <parent>
+    <groupId>com.exonum.binding</groupId>
+    <artifactId>exonum-java-binding-parent</artifactId>
+    <version>0.5-SNAPSHOT</version>
+    <relativePath>../pom.xml</relativePath>
+  </parent>
+
+  <artifactId>exonum-java-binding-packaging</artifactId>
+  <version>0.5-SNAPSHOT</version>
+  <packaging>pom</packaging>
```
Comment (dmitry-timofeev) [inline exonum-java-binding/packaging/pom.xml]: If it's POM, then we don't seem to need skipping JAR below (it's not bound in POM): https://maven.apache.org/ref/3.6.0/maven-core/default-bindings.html#Plugin_bindings_for_pom_packaging

File: exonum-java-binding/packaging/pom.xml [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+    <packaging.basedir>${project.parent.basedir}/core/rust/target/prepackage</packaging.basedir>
+    <packaging.rust>${project.parent.basedir}/core/rust</packaging.rust>
+    <packaging.java>${project.parent.basedir}/core/rust/target/debug/lib/java</packaging.java>
+  </properties>
+
+  <dependencies>
+    <dependency>
+      <groupId>com.exonum.binding</groupId>
+      <artifactId>exonum-java-binding-core</artifactId>
+      <version>0.5-SNAPSHOT</version>
```
Comment (dmitry-timofeev) [inline exonum-java-binding/packaging/pom.xml]: Here and elsewhere: ${project.version}

File: exonum-java-binding/packaging/pom.xml [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+      <groupId>com.exonum.binding</groupId>
+      <artifactId>exonum-time-oracle</artifactId>
+      <version>0.5-SNAPSHOT</version>
+      <type>pom</type>
+    </dependency>
+  </dependencies>
+
+  <build>
+    <plugins>
+      <plugin><!--Do not create Jar-->
```
Comment (dmitry-timofeev) [inline exonum-java-binding/packaging/pom.xml]: It is required only if this project uses 'jar' packaging which
binds jar and install plugins by default. But it is no longer the case (it's "pom" now).

I doubt, however, that the "pom" works, because it does not have "package" or "pre-package" phases that are used below. 

I'd add an issue to define a custom packaging for our use-case and *not* abuse the 'jar' packaging. It seems to be quite straightforward: https://stackoverflow.com/questions/1427722/how-do-i-create-a-new-packaging-type-for-maven

File: exonum-java-binding/packaging/pom.xml [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+                      <version>0.5-SNAPSHOT</version>
+                    </artifactItem>
+                  </artifactItems>
+                </configuration>
+              </execution>
+              <execution>
+                <id>copy-java-dependencies</id>
+                <phase>prepare-package</phase>
+                <goals>
+                  <goal>copy-dependencies</goal>
```
Comment (dmitry-timofeev) [inline exonum-java-binding/packaging/pom.xml]: > Goal that copies the project dependencies from the repository to a defined location.

Why then we copy core and time-oracle explicitly above?

File: exonum-java-binding/core/pom.xml [hunk 12]
```diff
@@ -342,6 +407,18 @@
...
             </goals>
           </execution>
+          <execution>
+            <id>copy-core-dependencies</id>
+            <phase>process-classes</phase>
+            <goals>
+              <goal>copy-dependencies</goal>
+            </goals>
+            <configuration>
+              <outputDirectory>${project.basedir}/rust/target/debug/lib/java</outputDirectory>
```
[the remark this answers is not in the document]
Comment (dmitry-timofeev) [reply exonum-java-binding/core/pom.xml]: The outputDirectories are duplicated :confused:

File: exonum-java-binding/core/pom.xml [hunk 13]
```diff
@@ -342,6 +374,35 @@
...
+                  <groupId>com.exonum.binding</groupId>
+                  <artifactId>exonum-java-binding-core</artifactId>
+                  <version>0.5-SNAPSHOT</version>
+                </artifactItem>
+              </artifactItems>
+            </configuration>
+          </execution>
+          <execution>
+            <id>copy-core-dependencies</id>
+            <phase>process-classes</phase>
```
Comment (dmitry-timofeev) [inline exonum-java-binding/core/pom.xml]: Why bound to 'process-classes'? How is it different from ^?

File: exonum-java-binding/core/pom.xml [hunk 14]
```diff
@@ -342,6 +374,35 @@
               <goal>build-classpath</goal>
             </goals>
           </execution>
+          <execution>
```
Comment (dmitry-timofeev) [inline exonum-java-binding/core/pom.xml]: Please document these executions (why they do what they do)

File: exonum-java-binding/core/rust/src/runtime/java_service_runtime.rs [hunk 15]
```diff
@@ -162,24 +180,40 @@ impl JavaServiceRuntime {
 
     /// Adds required EJB-related arguments to JVM configuration
     fn add_required_arguments(
-        args_builder: InitArgsBuilder,
-        ejb_config: EjbConfig,
+        mut args_builder: InitArgsBuilder,
+        runtime_config: &RuntimeConfig,
+        service_config: &ServiceConfig,
+        internal_config: InternalConfig,
```
[the remark this answers is not in the document]
Comment (skletsun) [reply exonum-java-binding/core/rust/src/runtime/java_service_runtime.rs]: With multiple services we get rid of ServiceConfig and InternalConfig at runtime creation.

File: exonum-java-binding/core/rust/ejb-app/TUTORIAL.md [hunk 16]
```diff
@@ -18,90 +19,81 @@ in the [user guide](https://exonum.com/doc/version/0.10/get-started/java-binding
...
 
-**Note:** using `cargo run` command requires working from the `ejb-app` directory.
-This would be fixed in the future EJB versions.
-
 #### Generate Template Config
+Use `--ejb-module-name` for fully qualified name of your service module.
 
 ```$sh
-$ cargo run -- generate-template testnet/common.toml
+$ ejb-app generate-template testnet/common.toml \
```
Comment (skletsun) [inline exonum-java-binding/core/rust/ejb-app/TUTORIAL.md]: Why we switch here from `cargo run` to `ejb-app` while we use `cargo run` in the QA service and cryptocurrency-demo starting scripts?

File: exonum-java-binding/packaging/pom.xml (hunk 11 above)
Comment (vitvakatu) [reply exonum-java-binding/packaging/pom.xml]: Because we use `type=pom` dependencies, and it seems like in this case `copy-dependencies` will not copy jars. Will be fixed later, with ECR-3039

File: exonum-java-binding/packaging/pom.xml (hunk 10 above)
Comment (vitvakatu) [reply exonum-java-binding/packaging/pom.xml]: Thanks, reverted to jar and made task ECR-3039

File: exonum-java-binding/core/rust/ejb-app/TUTORIAL.md (hunk 16 above)
Comment (vitvakatu) [reply exonum-java-binding/core/rust/ejb-app/TUTORIAL.md]: Because this is the documentation for end-user who unlikely will have cargo installed.

File: exonum-java-binding/packaging/pom.xml [hunk 17]
```diff
@@ -5,32 +5,37 @@
...
-  <version>0.5-SNAPSHOT</version>
-  <packaging>pom</packaging>
+  <version>0.6.0-SNAPSHOT</version>
+  <packaging>jar</packaging>
+
+  <name>Exonum Java Binding: Packaging</name>
+  <description>A fake module used for packaging EJB App</description>
 
   <properties>
+    <!--A directory used for storing immediate files while packaging the App-->
```
Comment (dmitry-timofeev) [inline exonum-java-binding/packaging/pom.xml]: I am not sure I understand 'immediate' here. What do you mean by that?

File: exonum-java-binding/packaging/pom.xml [hunk 18]
```diff
@@ -5,32 +5,37 @@
...
     <packaging.basedir>${project.parent.basedir}/core/rust/target/prepackage</packaging.basedir>
+    <!--A directory containing Java dependencies of the App-->
```
Comment (dmitry-timofeev) [inline exonum-java-binding/packaging/pom.xml]: This comment does not seem correct. Also, I also don't see the property being used. Maybe the other two properties shall use it?

File: exonum-java-binding/packaging/pom.xml [hunk 19]
```diff
@@ -5,32 +5,37 @@
...
     <packaging.rust>${project.parent.basedir}/core/rust</packaging.rust>
+    <!--A directory containing native dynamic libraries used by the App-->
```
Comment (dmitry-timofeev) [inline exonum-java-binding/packaging/pom.xml]: > native

The comment does not seem correct.

File: exonum-java-binding/packaging/pom.xml (hunk 17 above)
Comment (vitvakatu) [reply exonum-java-binding/packaging/pom.xml]: temporary, of course

File: exonum-java-binding/packaging/pom.xml (hunk 18 above)
Comment (vitvakatu) [reply exonum-java-binding/packaging/pom.xml]: 🤦‍♂️ I love IDEA revert changes algorithm, which just swapped the lines. This property is used in app_assembly.xml configuration, I can't define properties directly inside it.

File: exonum-java-binding/packaging/pom.xml [hunk 20]
```diff
@@ -16,11 +16,12 @@
   <description>A fake module used for packaging EJB App</description>
 
   <properties>
-    <!--A directory used for storing immediate files while packaging the App-->
+    <!--A directory used for storing temporary files while packaging the App-->
```
Comment (dmitry-timofeev) [inline exonum-java-binding/packaging/pom.xml]: Is there a documentation anywhere on what we place in `target/prepackage` and what in `target/debug`? Currently it's not clear why lib/java ends up in the latter; and (something) in the former. Also, why other things (like lib/java) are placed _outside_ of a *base*dir?

Comment (vitvakatu) [reply exonum-java-binding/packaging/pom.xml]: `target/debug` is filled by Cargo and therefore contains only libjava_bindings.so. In `target/prepackage` we temporarily store licenses and libstd.so. `target/debug/lib/native` and `target/debug/lib/java` are both required for running EJB App from `target` in our scripts. Their content will be copied to `lib/java` and `lib/native` inside App archive. 

As some kind of improvement we can store `libstd` inside `lib/native` as well, though it's not necessary. Also we can move `prepackage` to `target/debug` - it would be more consistent.

Comment (vitvakatu) [reply exonum-java-binding/packaging/pom.xml]: Made some improvements - now prepackage directory is replaced by the new one `target/debug/etc`. The whole directory structure is explained in package_app.sh script

Comment (dmitry-timofeev) [reply exonum-java-binding/packaging/pom.xml]: Are you saying

```
target/debug/prepackage/
                       - lib
                            + java
                            + rust
```
?

Comment (vitvakatu) [reply exonum-java-binding/packaging/pom.xml]: No, without prepackage.

File: exonum-java-binding/core/pom.xml [hunk 21]
```diff
@@ -335,14 +370,6 @@
           <outputFile>${project.build.directory}/ejb-core-classpath.txt</outputFile>
           <includeScope>runtime</includeScope>
         </configuration>
-        <executions>
```
Comment (dmitry-timofeev) [inline exonum-java-binding/core/pom.xml]: If the execution is removed, this plugin configuration does not make sense and must be removed too. The file seems unused.

File: exonum-java-binding/tests_profile [hunk 22]
```diff
@@ -49,8 +49,6 @@ export JAVA_LIB_DIR="$(find ${JAVA_HOME} -type f -name libjvm.\* | xargs -n1 dir
 echo "JAVA_LIB_DIR=${JAVA_LIB_DIR}"
 
 # Version of Rust used to build tests.
-## Stable works well unless you want benchmarks.
-# TODO: stable does not work well until ECR-1839 is resolved
```
Comment (dmitry-timofeev) [inline exonum-java-binding/tests_profile]: The bug is open. Also, there is https://jira.bf.local/browse/ECR-2982 :man_facepalming:

File: exonum-java-binding/core/pom.xml [hunk 23]
```diff
@@ -262,6 +270,33 @@
         </executions>
       </plugin>
 
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-resources-plugin</artifactId>
+        <executions>
+          <!--We need to copy libjava_bindings dynamic library to the <ejb-app-location>/lib/native directory-->
```
Comment (dmitry-timofeev) [inline exonum-java-binding/core/pom.xml]: We do that so that the application will work from target, correct? Do we do the same with _Java_ dependencies in `packaging` module for the same reason? If so, shall we move the latter to this module (and possibly simplify the `packaging` module)? Do I understand correctly, that if that's done, I could get an application working from target by just building core and its dependencies:

```sh
mvn package -pl exonum-java-binding-core -am
```

instead of (?) currently *required* (because of Java-dep-copying step in `packaging`):

```sh
mvn package # Build all modules
```

File: CONTRIBUTING.md [hunk 24]
```diff
@@ -23,6 +23,8 @@ You need to install the following dependencies:
   To install a specific Rust version, use `rustup install 1.32.0` command.
   * The [system dependencies](https://exonum.com/doc/version/0.10/get-started/install/) of Exonum. 
   You do _not_ need to manually fetch and compile Exonum.
+  * For automatic packaging of the EJB App you also need `CMake` installed in your system. Also on
```
Comment (bullet-tooth) [inline CONTRIBUTING.md]: "you also need ... Also you need ..."
I do not repeat twice twice 😄

Comment (bullet-tooth) [reply CONTRIBUTING.md]: Please add a link for `CMake` as It done for maven, JDK and rust

File: CONTRIBUTING.md [hunk 25]
```diff
@@ -23,6 +23,8 @@ You need to install the following dependencies:
   To install a specific Rust version, use `rustup install 1.32.0` command.
   * The [system dependencies](https://exonum.com/doc/version/0.10/get-started/install/) of Exonum. 
   You do _not_ need to manually fetch and compile Exonum.
+  * For automatic packaging of the EJB App you also need `CMake` installed in your system. Also on 
+  Mac you need a `coreutils` package installed.
```
Comment (bullet-tooth) [inline CONTRIBUTING.md]: please ad description/link

File: exonum-java-binding/core/pom.xml (hunk 21 above)
Comment (bullet-tooth) [reply exonum-java-binding/core/pom.xml]: I would fire an issue for replacing `maven-dependency-plugin` by `maven-assembly-plugin`

File: exonum-java-binding/core/pom.xml [hunk 26]
```diff
@@ -401,5 +428,55 @@
         </plugins>
       </build>
     </profile>
+
+    <profile>
+      <id>app-packaging</id>
```
Comment (bullet-tooth) [inline exonum-java-binding/core/pom.xml]:
```suggestion
      <id>package-app</id>
```

File: exonum-java-binding/core/pom.xml [hunk 27]
```diff
@@ -401,5 +428,55 @@
...
+      <build>
+        <plugins>
+          <plugin>
+            <groupId>org.codehaus.mojo</groupId>
+            <artifactId>exec-maven-plugin</artifactId>
+            <configuration>
+              <executable>cargo</executable>
+              <workingDirectory>${project.basedir}/rust/</workingDirectory>
```
Comment (bullet-tooth) [inline exonum-java-binding/core/pom.xml]: do we need tailing `/`?

File: exonum-java-binding/core/rust/ejb-app/TUTORIAL.md [hunk 28]
```diff
@@ -18,90 +19,81 @@ in the [user guide](https://exonum.com/doc/version/0.10/get-started/java-binding
...
 ```bash
 JAVA_HOME="${JAVA_HOME:-$(java -XshowSettings:properties -version 2>&1 > /dev/null | grep 'java.home' | awk '{print $3}')}"
 LIBJVM_PATH="$(find ${JAVA_HOME} -type f -name libjvm.* | xargs -n1 dirname)"
 
-RUST_LIB_PATH="$(rustup run 1.32.0 rustc --print sysroot)/lib"
-
-export EJB_LIBPATH="${EJB_ROOT}/core/rust/target/debug"
-
-export LD_LIBRARY_PATH="${LIBJVM_PATH}:${RUST_LIB_PATH}:${EJB_LIBPATH}"
+export LD_LIBRARY_PATH="${LIBJVM_PATH}"
```
Comment (bullet-tooth) [inline exonum-java-binding/core/rust/ejb-app/TUTORIAL.md]: let's remove LD_LIBRARY_PATH and export LIBJVM_PATH

File: exonum-java-binding/core/rust/ejb-app/TUTORIAL.md [hunk 29]
```diff
@@ -18,90 +19,81 @@ in the [user guide](https://exonum.com/doc/version/0.10/get-started/java-binding
...
 
 You may package your service in an Uber JAR using 
 the [Maven Shade Plugin](https://maven.apache.org/plugins/maven-shade-plugin/index.html)
-and pass a path to the service artefact during application configuration as `--ejb-classpath`
-parameter. Alternatively, you may assemble a classpath that includes the service and all of 
+and pass a path to the service artefact during application configuration as `--ejb-service-classpath`
+parameter. Alternatively, you may assemble a classpath that includes the path to service and all of 
 its dependencies and pass it instead.
 
+The service must use `provided` scope for all Exonum dependencies because they are included in the application.
```
Comment (bullet-tooth) [inline exonum-java-binding/core/rust/ejb-app/TUTORIAL.md]: I would mention these dependencies in a list.
Also we use assembly plugin, not shade plugin. But this change doesn't belong to this PR

File: exonum-java-binding/core/rust/ejb-app/TUTORIAL.md [hunk 30]
```diff
@@ -18,90 +19,81 @@ in the [user guide](https://exonum.com/doc/version/0.10/get-started/java-binding
...
 ### Step 2. Generate Node Configuration
 
 EJB App configuration is pretty similar to configuration of any other Exonum service,
 with a few additional parameters.
 
-**Note:** using `cargo run` command requires working from the `ejb-app` directory.
-This would be fixed in the future EJB versions.
-
 #### Generate Template Config
+Use `--ejb-module-name` for fully qualified name of your service module.
```
Comment (bullet-tooth) [inline exonum-java-binding/core/rust/ejb-app/TUTORIAL.md]: Should it be a full Java class name?

File: exonum-java-binding/core/rust/ejb-app/start_cryptocurrency_node.sh [hunk 31]
```diff
@@ -37,41 +37,34 @@ echo "PROJ_ROOT=${EJB_ROOT}"
...
-EJB_LOG_CONFIG_PATH="${EJB_APP_DIR}/log4j2.xml"
+EJB_LOG_CONFIG_PATH="${EJB_APP_DIR}/log4j-fallback.xml"
 
-EJB_LIBPATH="${EJB_ROOT}/core/rust/target/debug"
-echo "EJB_LIBPATH=${EJB_LIBPATH}"
-RUST_LIB_DIR="$(rustup run 1.32.0 rustc --print sysroot)/lib"
-echo "RUST_LIB_DIR=${RUST_LIB_DIR}"
-
-export LD_LIBRARY_PATH="$EJB_LIBPATH:$RUST_LIB_DIR:$JVM_LIB_PATH"
+export LD_LIBRARY_PATH="$JVM_LIB_PATH"
```
Comment (bullet-tooth) [inline exonum-java-binding/core/rust/ejb-app/start_cryptocurrency_node.sh]: can be removed

File: exonum-java-binding/core/rust/ejb-app/start_qa_node.sh [hunk 32]
```diff
@@ -36,17 +36,11 @@ echo "PROJ_ROOT=${EJB_ROOT}"
...
-EJB_CLASSPATH="$(cat ${EJB_ROOT}/${CORE_TXT}):$(cat ${EJB_ROOT}/${QA_SERVICE_TXT})"
-EJB_CLASSPATH="${EJB_CLASSPATH}:${EJB_ROOT}/core/target/classes"
+EJB_CLASSPATH="$(cat ${EJB_ROOT}/${QA_SERVICE_TXT})"
 EJB_CLASSPATH="${EJB_CLASSPATH}:${EJB_ROOT}/qa-service/target/classes"
 echo "EJB_CLASSPATH=${EJB_CLASSPATH}"
 
-EJB_LIBPATH="${EJB_ROOT}/core/rust/target/debug"
-echo "EJB_LIBPATH=${EJB_LIBPATH}"
-export RUST_LIB_DIR="$(rustup run 1.32.0 rustc --print sysroot)/lib"
-export LD_LIBRARY_PATH="$LD_LIBRARY_PATH":"$EJB_LIBPATH":"$RUST_LIB_DIR"
```
Comment (bullet-tooth) [inline exonum-java-binding/core/rust/ejb-app/start_qa_node.sh]: can be removed

File: exonum-java-binding/core/rust/ejb-app/start_qa_node.sh [hunk 33]
```diff
@@ -69,27 +63,23 @@ do
 done
 
 header "GENERATE COMMON CONFIG"
-cargo run -- generate-template --validators-count $node_count testnet/common.toml
+cargo run -- generate-template --validators-count $node_count testnet/common.toml \
```
Comment (bullet-tooth) [inline exonum-java-binding/core/rust/ejb-app/start_qa_node.sh]: why does it use `cargo`, not `ejb-app`?

File: exonum-java-binding/core/pom.xml (hunk 21 above)
Comment (dmitry-timofeev) [reply exonum-java-binding/core/pom.xml]: What do you mean by replacing and what do you expect from assembly-plugin? There used to be a configuration to generate a classpath file that was used in scripts launching QA/cryptocurrency services. As the app creates a classpath from JARs in "lib/java", I don't see if anything is needed here (except copying the JARs).

File: exonum-java-binding/core/rust/ejb-app/TUTORIAL.md (hunk 29 above)
Comment (dmitry-timofeev) [reply exonum-java-binding/core/rust/ejb-app/TUTORIAL.md]: Do you think the whole list is needed here (it is a tutorial on how to launch a service you have already developed, including project configuration)? Probably, it is more appropriate in the docs: https://github.com/exonum/exonum-doc/pull/301/files#diff-def2f510ddeb1ee71c53f683914e41c6R496

File: exonum-java-binding/core/rust/ejb-app/TUTORIAL.md (hunk 30 above)
Comment (dmitry-timofeev) [reply exonum-java-binding/core/rust/ejb-app/TUTORIAL.md]: Isn't "fully qualified name" clear enough? Or are you saying that it might be unclear that service module is a Java class?

Comment (dmitry-timofeev) [reply exonum-java-binding/core/rust/ejb-app/TUTORIAL.md]: Anyway, this documentation won't even survive till the next release because it is configured differently in "multiple-services"

File: exonum-java-binding/core/rust/ejb-app/start_cryptocurrency_node.sh (hunk 31 above)
Comment (dmitry-timofeev) [reply exonum-java-binding/core/rust/ejb-app/start_cryptocurrency_node.sh]: What exactly? @vitvakatu Isn't this var required to locate libjvm?

File: exonum-java-binding/core/rust/ejb-app/start_cryptocurrency_node.sh [hunk 34]
```diff
@@ -37,41 +37,34 @@ echo "PROJ_ROOT=${EJB_ROOT}"
...
 echo "Final LD_LIBRARY_PATH=${LD_LIBRARY_PATH}"
```
Comment (dmitry-timofeev) [inline exonum-java-binding/core/rust/ejb-app/start_cryptocurrency_node.sh]: Now it makes sense to say just "LD_LIBRARY_PATH" :upside_down_face:

File: exonum-java-binding/core/pom.xml (hunk 23 above)
Comment (vitvakatu) [reply exonum-java-binding/core/pom.xml]: We also need to copy time-oracle classes, so it can't be done in Core module. That's why we actually created Packaging module

File: exonum-java-binding/core/pom.xml (hunk 21 above)
Comment (vitvakatu) [reply exonum-java-binding/core/pom.xml]: Agreed, removed.

File: CONTRIBUTING.md (hunk 24 above)
Comment (vitvakatu) [reply CONTRIBUTING.md]: Fixed

File: CONTRIBUTING.md (hunk 25 above)
Comment (vitvakatu) [reply CONTRIBUTING.md]: fixed

File: exonum-java-binding/core/rust/ejb-app/TUTORIAL.md (hunk 28 above)
Comment (vitvakatu) [reply exonum-java-binding/core/rust/ejb-app/TUTORIAL.md]: It's impossible, we need to export LD_LIBRARY_PATH

File: exonum-java-binding/core/rust/ejb-app/start_qa_node.sh (hunk 33 above)
Comment (vitvakatu) [reply exonum-java-binding/core/rust/ejb-app/start_qa_node.sh]: Because you can't guarantee the binary ejb-app in `target` directory would be valid (linked against right libraries). When running integration tests, we build a special version of libjava_bindings, which can't be used with the `ejb-app` binary build for packaging, for example. So we run cargo just to ensure if the right library is provided and recompile otherwise. After running this script for the first time you can, however, run the binary directly from target.

File: exonum-java-binding/core/rust/ejb-app/start_cryptocurrency_node.sh (hunk 31 above)
Comment (vitvakatu) [reply exonum-java-binding/core/rust/ejb-app/start_cryptocurrency_node.sh]: Yep, exactly

File: exonum-java-binding/core/pom.xml (hunk 23 above)
Comment (dmitry-timofeev) [reply exonum-java-binding/core/pom.xml]: I see, that makes sense.

File: exonum-java-binding/core/rust/ejb-app/start_qa_node.sh (hunk 33 above)
Comment (dmitry-timofeev) [reply exonum-java-binding/core/rust/ejb-app/start_qa_node.sh]: > running integration tests, we build a special version of libjava_bindings

Rust unit tests (no-resource-manager)? Otherwise I don't understand.

---

Shall @dip56 create a separate script which uses an **installed** application? And this be kept for development purposes?

Comment (vitvakatu) [reply exonum-java-binding/core/rust/ejb-app/start_qa_node.sh]: Oh, yes, unit tests without resource manager. Created ECR-3063 for end-users support. These scripts will be used for developing purposes only.

**Comment by dmitry-timofeev** [comment]
@vitvakatu Do we squash it or merge as is?

**Comment by vitvakatu** [comment]
@dmitry-timofeev as you wish, I'm okay with squashing
